codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,210 @@
+        // here since the Writer could call updateStreamableBytes, which may alter the queue.
+        do {
+            // Remove the head of the queue.
+            State state = queue.remove();
+            state.enqueued = false;
+
+            // Allocate as much data as we can for this stream.
+            int chunk = min(chunkSize, min(maxBytes, state.streamableBytes));
+            maxBytes -= chunk;
+
+            // Write the allocated bytes and enqueue as necessary.
+            state.write(chunk, writer);
+        } while (maxBytes > 0 && !queue.isEmpty());
+
+        return totalStreamableBytes > 0;
+    }
+
+    private State state(Http2Stream stream) {
+        return checkNotNull(stream, "stream").getProperty(stateKey);
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class State {
+        final Http2Stream stream;
+        int streamableBytes;
+        boolean enqueued;
+        boolean wasEnqueued;
+
+        State(Http2Stream stream) {
+            this.stream = stream;
+        }
+
+        void updateStreamableBytes(int newStreamableBytes, boolean hasFrame) {
+            assert hasFrame || newStreamableBytes == 0;
+
+            int delta = newStreamableBytes - streamableBytes;
+            if (delta != 0) {
+                streamableBytes = newStreamableBytes;
+                totalStreamableBytes += delta;
+            }
+            if (hasFrame) {
+                // It's not in the queue but has data to send, add it.
+                addToQueue();
+            }
+        }
+
+        /**
+         * Write any allocated bytes for the given stream and updates the streamable bytes,
+         * assuming all of the bytes will be written.
+         */
+        void write(int numBytes, Writer writer) throws Http2Exception {
+            // Update the streamable bytes, assuming that all the bytes will be written.
+            int newStreamableBytes = streamableBytes - numBytes;
+            updateStreamableBytes(newStreamableBytes, newStreamableBytes > 0);
+
+            try {
+                // Write the allocated bytes.
+                writer.write(stream, numBytes);
+            } catch (Throwable t) {
+                throw connectionError(INTERNAL_ERROR, t, "byte distribution write error");
+            }
+        }
+
+        void addToQueue() {
+            if (!enqueued) {
+                if (streamableBytes == 0) {
+                    // Add empty frames to the empty frame queue.
+                    emptyFrameQueue.addLast(this);
+                } else if (!wasEnqueued) {
+                    // Add to the head of the list the first time this stream is enqueued.
+                    queue.addFirst(this);
+                } else {
+                    queue.addLast(this);
+                }
+                enqueued = true;
+                wasEnqueued = true;
```

Comment:
if the item is added to `emptyFrameQueue` but then more frames are added we will remove it from `emptyFrameQueue` and add it to the end of `queue`. Should we just add this element to the front of `queue`? Should we move this `wasEnqueued = true` when we first add to `wasEnqueued` (and maybe rename)?